Sources/Testing/SourceAttribution/Expression.swift

```diff
@@ -0,0 +1,261 @@
+          return argument.value.sourceCode
+        }.joined(separator: ", ")
+      if let value {
+        return "\(value.sourceCode).\(functionName)(\(argumentList))"
+      }
+      return "\(functionName)(\(argumentList))"
+    case let .propertyAccess(value, keyPath):
+      return "\(value.sourceCode).\(keyPath.sourceCode)"
+    }
+  }
+
+  /// A description, captured using ``String/init(describingForTest:)``, of
+  /// the runtime value of this subexpression.
+  ///
+  /// If the runtime value of this subexpression has not been evaluated, the
+  /// value of this property is `nil`.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var runtimeValueDescription: String?
+
+  /// The fully-qualified name of the type of this subexpression's runtime
+  /// value.
+  ///
+  /// If the runtime value of this subexpression has not been evaluated, the
+  /// value of this property is `nil`.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var fullyQualifiedTypeNameOfRuntimeValue: String?
+
+  /// Copy this instance and capture the runtime values corresponding to its
+  /// subexpressions.
+  ///
+  /// - Parameters:
+  ///   - firstValue: The first captured runtime value.
+  ///   - additionalValues: Any additional captured runtime values after the
+  ///     first.
+  ///
+  /// - Returns: A copy of `self` with information about the specified runtime
+  ///   values captured for future use.
+  func capturingRuntimeValues<each T>(_ firstValue: (some Any)?, _ additionalValues: repeat (each T)?) -> Self {
+    var result = self
+
+    // Convert the variadic generic argument list to an array.
+    var additionalValuesArray = [Any?]()
+    repeat additionalValuesArray.append(each additionalValues)
+
+    switch kind {
+    case .generic:
+      if let firstValue {
+        result.runtimeValueDescription = String(describingForTest: firstValue)
+        result.fullyQualifiedTypeNameOfRuntimeValue = _typeName(type(of: firstValue as Any), qualified: true)
+      } else {
+        result.runtimeValueDescription = nil
+        result.fullyQualifiedTypeNameOfRuntimeValue = nil
+      }
+    case let .binaryOperation(lhsExpr, op, rhsExpr):
+      result.kind = .binaryOperation(
+        lhs: lhsExpr.capturingRuntimeValues(firstValue),
+        operator: op,
+        rhs: rhsExpr.capturingRuntimeValues(additionalValuesArray.first ?? nil)
+      )
+    case let .functionCall(value, functionName, arguments):
+      result.kind = .functionCall(
+        value: value?.capturingRuntimeValues(firstValue),
+        functionName: functionName,
+        arguments: zip(arguments, additionalValuesArray).map { argument, value in
+          .init(label: argument.label, value: argument.value.capturingRuntimeValues(value))
+        }
+      )
+    case let .propertyAccess(value, keyPath):
+      result.kind = .propertyAccess(
+        value: value.capturingRuntimeValues(firstValue),
+        keyPath: keyPath.capturingRuntimeValues(additionalValuesArray.first ?? nil)
```

Comment:
This is probably a pre-existing concern, but I wonder about error handling here, if the number of runtime values does not match the number of source code sub-expressions. What do you think about at least having a "FIXME" comment or similar noting that it could happen and may need to be addressed at some point? My only worry is tha usage of `zip` (for example) could lead to something getting silently dropped

Reply:
I don't know what you're specifically concerned about so I am unsure what I'd put in a FIXME. Can you clarify?

Reply:
Just the possibility that the number of variadic generic parameters passed to this function might not equal the number of subexpressions. If they mismatch, the `zip` function will use the shorter value, and could cause some data to be lost

Reply:
The number of arguments vs. values will be equal in that case, because the source code will just look like:

```swift
f(a, b, c)
```

And the set of values will be `[a, b, c]`. We have no knowledge of the _signature_ of `f()` in this spot, so we have no knowledge of potential variadic sequences. We just see the function call expression.

Reply:
If we come up with a concrete problem, then yes let's document it and try to guard against it. :)